momento-sdk/src/main/java/momento/sdk/TopicClient.java

```diff
@@ -0,0 +1,104 @@
+    logger.info("Creating Momento Topic Client");
+    logger.debug("Cache endpoint: " + credentialProvider.getCacheEndpoint());
+    logger.debug("Control endpoint: " + credentialProvider.getControlEndpoint());
+  }
+
+  /**
+   * Constructs a TopicClient.
+   *
+   * @param credentialProvider Provider for the credentials required to connect to Momento.
+   * @param configuration Configuration object containing all tunable client settings.
+   * @return TopicClient
+   */
+  public static TopicClient create(
+      @Nonnull CredentialProvider credentialProvider, @Nonnull Configuration configuration) {
+    return create(credentialProvider, configuration);
+  }
+
+  /**
+   * Creates a TopicClient builder.
+   *
+   * @param credentialProvider Provider for the credentials required to connect to Momento.
+   * @param configuration Configuration object containing all tunable client settings.
+   * @return The builder.
+   */
+  public static TopicClientBuilder builder(
+      CredentialProvider credentialProvider, Configuration configuration) {
+    return new TopicClientBuilder(credentialProvider, configuration);
+  }
+
+  /**
+   * Publish a message to a topic with provided topic name in a cache with provided cache name.
+   *
+   * @param cacheName The name of the cache where topic resides.
+   * @param topicName The name of the topic.
+   * @param message The message to be published.
+   * @return A future containing the result of the topic publish: {@link
+   *     TopicPublishResponse.Success} or {@link TopicPublishResponse.Error}.
+   */
+  public CompletableFuture<TopicPublishResponse> publish(
+      String cacheName, String topicName, String message) {
+    return scsTopicClient.publish(cacheName, topicName, message);
+  }
+
+  /**
+   * Publish a message to a topic with provided topic name in a cache with provided cache name.
+   *
+   * @param cacheName The name of the cache where topic resides.
+   * @param topicName The name of the topic.
+   * @param message The message to be published.
+   * @return A future containing the result of the topic publish: {@link
+   *     TopicPublishResponse.Success} or {@link TopicPublishResponse.Error}.
+   */
+  public CompletableFuture<TopicPublishResponse> publish(
+      String cacheName, String topicName, byte[] message) {
+    return scsTopicClient.publish(cacheName, topicName, message);
+  }
+
+  /**
+   * Subscribe to a topic with provided topic name in a cache with provided cache name.
+   *
+   * @param cacheName The name of the cache where topic resides.
+   * @param topicName The name of the topic.
+   * @param options The options for handling the subscription callbacks. See {@link
+   *     ISubscribeCallOptions}.
+   * @return A future containing the result of the topic subscribe: {@link
+   *     TopicSubscribeResponse.Subscription} or {@link TopicSubscribeResponse.Error}.
+   */
+  public CompletableFuture<TopicSubscribeResponse> subscribe(
```

Comment:
100% agreed on `subscribeResponse.orElseThrow(() -> new RuntimeException("Subscription failed: " + subscribeResponse));` its more idiomatic and was released in Java 8; so its also quite well accepted within the community. The only argument that folks might have is that we do require casting for our cache methods, so users of both topic and cache clients would have a dual experience. Although with caches we have established a sealed class pattern with different possibilities that don't apply here, so we can live with that. I'd still +1 on the Optional as it looks way neater.

Reply:
I am a complete java :noob: so if what I am saying doesnt make sense can ignore it, but that looks like to me that `.orElseThrow` violates our pattern set of explicitly _not_ throwing exceptions. To me its weird that none of our methods throw exceptions except this one? Or maybe `.orElseThrow` is doing something that I am not expecting?

Reply:
To clarify, why would we throw an exception rather that returning the error response?

Reply:
Yeah fair that's what I said that it's a valid argument that we are going to deviate from some of the established patterns. I think we could still use optional and use `orElse(return Error)` instead of `orElseThrow` to your concern Matt.

Reply:
Ok cool that approach seems more consistent to me, I like it 👍

Reply:
> I am a complete java :noob: so if what I am saying doesnt make sense can ignore it, but that looks like to me that .orElseThrow violates our pattern set of explicitly not throwing exceptions. 

in this case the user is explicitly telling us they want to throw so I think it would be okay. but we could totally support both `orElse` and `orElseThrow`, they're not exclusive.

> The only argument that folks might have is that we do require casting for our cache methods, so users of both topic and cache clients would have a dual experience. 

I don't think this is necessarily true? If we have any "best practice"/example code for our existing APIs that requires explicit casting, then I would want to provide alternatives for that ASAP. In the switch statements I think it's fine, when the language itself is doing like, "smart casting", but if there is anywhere that a user can't use our SDK with out doing an explicit cast via `()` then I would like to dig into that so please let me know.